torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -352,18 +392,26 @@ def _init_from_local_shards(
             gathered_metadatas = [local_sharded_tensor_metadata]
 
         global_sharded_tensor_metadata = build_global_metadata(gathered_metadatas)
+        tensor_properties = global_sharded_tensor_metadata.tensor_properties
 
         # STEP 3: Validation done, create the actual ShardedTensor and populate fields
         # prepare initialization
-        sharded_tensor = cls.__new__(cls)
+        spec = shard_spec._infer_sharding_spec_from_shards_metadata(
+            global_sharded_tensor_metadata.shards_metadata
+        )
+        sharded_tensor = cls.__new__(cls,
+                                     spec,
+                                     global_sharded_tensor_metadata.size,
+                                     dtype=tensor_properties.dtype,
+                                     layout=tensor_properties.layout,
+                                     pin_memory=tensor_properties.pin_memory,
+                                     requires_grad=tensor_properties.requires_grad)
         sharded_tensor._prepare_init(process_group=process_group, init_rrefs=init_rrefs)
 
         # add to metadata and local_shards
         sharded_tensor._metadata = global_sharded_tensor_metadata
         sharded_tensor._local_shards = local_shards
```

Comment:
> Maybe we don't need to store this field statically, the ground truth metadata can be located inside the List[Shard], when user call ShardedTensor.metdata(), we can dynamically do an all_gather and return a ShardedTensorMetadata. What do you think about this?

A bit worried about this due to the additional all_gather call since I'm assuming there might be already usages of `.metadata()`. Btw, can't we move `_metadata` to the base class? We should be able to create the metadata from the sharding_spec and sizes/dtype etc. right?

Maybe we can do the reverse and store ground truth as ShardedTensorMetadata and dynamically build `List[Shard]` when users call `local_shards()` since that can be built without any comms.

Reply:
Agreed that can move `_metadata` to the base class, the trickly part is that `ShardedTensorMetadata` contains `ShardMetadata`, which is something that extension user might not want to use (i.e. they might want their own custom shardmetadata). In this case they will have two source of truths on sharding layout (1. our ShardMetadata layout built by `sharding_spec.build_metadata` 2. their custom shardmetadata layout). I guess they can silently ignore the first one and use their own layout description in this case?

Reply:
Not sure I followed this completely, the way the extension interface works is that `sharding_spec.build_metadata` returns `ShardedTensorMetadata`, so the extension has to implement a way to return this metadata to us. So we can always call that method for metadata.

Now, the user can do their own thing, but then thats on their plate to deal with two source of truths. Ideally, they should just use `ShardedTensorMetadata` to avoid this.

Reply:
Yep, we call `.metadata()` in the sharded ops a lot, for example, chunk. So it will be nice that if we can avoid the communication here.